### PR TITLE
Bug 1700946 - Update BackfillRecord after soft launch

### PR DESCRIPTION
**Pull request by airimovici**
Bugzilla bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1700946

### CODE REVIEW

File: treeherder/perf/models.py [hunk 1]
```diff
@@ -580,6 +580,15 @@ class BackfillRecord(models.Model):
     def id(self):
         return self.alert
 
+    def get_pushes_in_context_range(self):
```
Comment (ionutgoldan) [inline treeherder/perf/models.py:583]: We should also use this method in `perf.email.BackfillReportContent.__fetch_border_pushes()`.

**Review by ionutgoldan — CHANGES_REQUESTED** [review]

File: treeherder/perf/auto_perf_sheriffing/perf_sheriff_bot.py [hunk 2]
```diff
@@ -132,10 +132,11 @@ def _backfill_record(self, record: BackfillRecord, left: int) -> Tuple[int, int]
                     break
                 try:
                     using_job_id = data_point['job_id']
+                    logger.info(f'Trying to backfill using job_id {using_job_id}')
                     self.backfill_tool.backfill_job(using_job_id)
                     left, consumed = left - 1, consumed + 1
                 except (KeyError, CannotBackfill, Exception) as ex:
-                    logger.debug(f'Failed to backfill record {record.alert.id}: {ex}')
+                    logger.info(f'Failed to backfill record {record.alert.id}: {ex}')
```
Comment (ionutgoldan) [inline treeherder/perf/auto_perf_sheriffing/perf_sheriff_bot.py:139]: Shouldn't we keep this at debug level?

File: tests/perfalert/test_auto_perf_sheriffing/test_secretary_tool.py [hunk 3]
```diff
@@ -22,11 +22,18 @@ def get_middle_index(successful_jobs):
     return index_in_range
 
 
+def set_record_job_type(record, job_type_id=JOB_TYPE_ID):
+    record.job_type = JobType.objects.filter(id=job_type_id).first()
```
Comment (ionutgoldan) [inline tests/perfalert/test_auto_perf_sheriffing/test_secretary_tool.py:26]: You don't need to fetch the `JobType` instance. `job_type_id` is accessible as attribute of `record`.

File: tests/perfalert/test_auto_perf_sheriffing/test_secretary_tool.py [hunk 4]
```diff
@@ -22,11 +22,18 @@ def get_middle_index(successful_jobs):
...
+    record.save()
+
+
 @pytest.fixture
 def record_backfilled(test_perf_alert, record_context_sample):
     report = BackfillReport.objects.create(summary=test_perf_alert.summary)
+    job_type = JobType.objects.filter(id=JOB_TYPE_ID).first()
     record = BackfillRecord.objects.create(
         alert=test_perf_alert,
+        job_type=job_type,
```
Comment (ionutgoldan) [inline tests/perfalert/test_auto_perf_sheriffing/test_secretary_tool.py:36]: Same here.

File: treeherder/perf/auto_perf_sheriffing/outcome_checker.py [hunk 5]
```diff
@@ -37,11 +20,8 @@ def __init__(self):
         pass
 
     def check(self, record: BackfillRecord) -> OutcomeStatus:
-        # TODO: get job_type from record when soft launch lands ---> job_type = record.job_type
-        job_type = get_job_type(record)
-        from_time, to_time = self._get_push_timestamp_range(record.get_context())
-        repository_id = record.alert.summary.repository.id
-        pushes_in_range = self._get_pushes_in_range(from_time, to_time, repository_id)
+        job_type = record.job_type
```
Comment (ionutgoldan) [inline treeherder/perf/auto_perf_sheriffing/outcome_checker.py:23]: I think we should raise a `ValueError` is job type is null.

File: treeherder/perf/auto_perf_sheriffing/outcome_checker.py [hunk 6]
```diff
@@ -37,11 +20,8 @@ def __init__(self):
...
+        pushes_in_range = record.get_pushes_in_context_range()
 
         for push in pushes_in_range:
             # make sure it has at least one successful job of job type
```
Comment (ionutgoldan) [inline treeherder/perf/auto_perf_sheriffing/outcome_checker.py:27]: We could make this code even more readable.

File: treeherder/perf/auto_perf_sheriffing/outcome_checker.py [hunk 7]
```diff
@@ -37,11 +20,8 @@ def __init__(self):
         pass
 
     def check(self, record: BackfillRecord) -> OutcomeStatus:
```
Comment (ionutgoldan) [inline treeherder/perf/auto_perf_sheriffing/outcome_checker.py:22]:
```suggestion
    def check(self, record: BackfillRecord) -> OutcomeStatus:
        of_type = record.job_type
        with_unknown_results = 'unknown'
        pushes_in_range = record.get_pushes_in_context_range()

        for push in pushes_in_range:
            # make sure it has at least one successful job of job type
            if push.total_successful_jobs(of_type) == 0:
                # either (at least) one job is in progress or it failed
                if push.total_unfinished_jobs(of_type, with_unknown_results) > 0:
                    return OutcomeStatus.IN_PROGRESS
                else:
                    return OutcomeStatus.FAILED

        return OutcomeStatus.SUCCESSFUL
```

Comment (ionutgoldan) [reply treeherder/perf/auto_perf_sheriffing/outcome_checker.py:22]: What do you think of these cosmetics?

**Review by ionutgoldan — COMMENTED** [review]
Left some extra comments.

File: treeherder/perf/models.py [hunk 8]
```diff
@@ -611,6 +611,27 @@ class BackfillRecord(models.Model):
...
+        """
+        Provides border(first and last) information from context based on the property
+        """
+        context = self.get_context()
+        from_info = context[0][context_property]
+        to_info = context[-1][context_property]
+
+        return from_info, to_info
+
+    def get_pushes_in_context_range(self):
```
Comment (ionutgoldan) [inline treeherder/perf/models.py:628]: Please provide appropriate type hint for function' s output. It'll help navigate through the code.

File: treeherder/perf/auto_perf_sheriffing/outcome_checker.py (hunk 7 above)
Comment (airimovici) [reply treeherder/perf/auto_perf_sheriffing/outcome_checker.py:22]: Thanks for the suggestions. I added a method to the Push model for getting the total number of jobs of a type with a particular result(`success` or `unknown`).

File: tests/perfalert/test_auto_perf_sheriffing/test_perf_sheriff_bot.py [hunk 9]
```diff
@@ -71,7 +71,7 @@ def test_no_email_is_sent_if_context_is_too_corrupt_to_be_actionable(
         )
         sheriff_bot.sheriff(since=EPOCH, frameworks=['raptor', 'talos'], repositories=['autoland'])
 
-        assert notify_client_mock.email.call_count == 0
+        assert notify_client_mock.email.call_count == 1
```
Comment (ionutgoldan) [inline tests/perfalert/test_auto_perf_sheriffing/test_perf_sheriff_bot.py:74]: Oh, we need to rename the test.

**Review by ionutgoldan — CHANGES_REQUESTED** [review]

**Review by ionutgoldan — APPROVED** [review]

**Pull request merged**